### PR TITLE
add out variant of linear

### PR DESCRIPTION
**Pull request by makslevental**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61824
* #61821
* #61802
* __->__ #61801
* #61800

resubmitting because the last one was unrecoverable due to making changes incorrectly in the stack

Differential Revision: [D29812510](https://our.internmc.facebook.com/intern/diff/D29812510)

### CODE REVIEW

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu [hunk 1]
```diff
@@ -727,9 +727,9 @@ namespace {
 } // namespace
 
   Tensor& adaptive_avg_pool2d_out_cuda(
-    Tensor& output,
     const Tensor& input,
-    IntArrayRef output_size)
+    IntArrayRef output_size,
+    Tensor& output)
```
Comment (desertfire) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:732]: Forgot the include your previous fix in native_functions.yaml? Can we make this as a separate PR since you have mastered ghstack? It will make the review and triage easier if anything goes wrong.

Comment (makslevental) [reply aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu]: @desertfire https://github.com/pytorch/pytorch/pull/61851

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812510).

File: test/cpp/api/operations.cpp [hunk 2]
```diff
@@ -55,3 +55,33 @@ TEST_F(OperationTest, Cross) {
     ASSERT_TRUE(out.allclose(exp));
   }
 }
+
+TEST_F(OperationTest, Linear_out) {
```
Comment (desertfire) [inline test/cpp/api/operations.cpp:59]: I remember Nick posted some comments about this test, mainly on the integer type. Can we use an approach similar to `TEST_F(OperationTest, Cross)` where we use `torch::rand` to generate the input and use loops to compute the expected result?

Comment (makslevental) [reply test/cpp/api/operations.cpp:59]: @desertfire nick added a comment about using `OpInfo` instead of this approach. I agree it's a better approach but I would like to land this one and then submit the `OpInfo` in a separate PR.

**Comment by makslevental** [comment]
@desertfire , @hlu1 , @eellison and @Krovatkin  any blockers here?

**Review by desertfire — APPROVED** [review]

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812510).
